butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java

```diff
@@ -89,6 +89,7 @@ private void setEnvironment(String[] arguments) {
                 System.exit(1);
             }
         }
+        LogFileDefiner.updateLog4jConfigWithLogFile();
```

Comment:
This is absolutely not the right place to for this line of code. I realise now. 
In case of Logback, the file name was retrieved when the framework initialises by calling [`getPropertyValue()`](https://github.com/paypal/butterfly/blob/d7c9819de9e7516d32950b2348c71b9d424d691a/butterfly-cli/src/main/java/com/paypal/butterfly/cli/logging/LogFileDefiner.java#L51). 

In Log4j2 I could not find a way to set the log file name during initialisation by calling this function. So, I used system property. The system property value should be update before any calls to Logger. So, I think, this method should be called [here](https://github.com/paypal/butterfly/blob/d7c9819de9e7516d32950b2348c71b9d424d691a/butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java#L46) right after we set the `butterflyHome`. Later on, when we update the log file name, we can updated the system property [here](https://github.com/paypal/butterfly/blob/d7c9819de9e7516d32950b2348c71b9d424d691a/butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java#L84) so that new file name is used by logger. What do you think? This may also depend on what we decide to do with the `LogFileDefiner` class which you mentioned below.

Reply:
Ok, we got a conclusion about the situation with `LogFileDefiner` (see comment below). Now let's go ahead and address this too. Also, is setting the System property really the only way to set the log file? Isn't there a way to do it using Log4j 2 API?

Reply:
Let me do some work on that. I will get back soon.

Reply:
I looked more into the documentations. Here is what I found so far.

Log4j2 doesn't have any API to directly set the value of `filename` dynamically. I checked [FileAppender](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/appender/FileAppender.html) and [RandomAccessFileAppender](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/appender/RandomAccessFileAppender.html). So, we have two options- use property substitution or programmatically configure Log4j2. 

There are different property substitution. Please read [here](https://logging.apache.org/log4j/2.x/manual/configuration.html#PropertySubstitution) and [here](https://logging.apache.org/log4j/2.x/manual/lookups.html). I have used `system` lookup. If we don't want to use `System` property, we can create a [lookup plugin](https://logging.apache.org/log4j/2.x/manual/plugins.html#Lookups) and use it to get the `logFile` name dynamically. The difference between this approach and using `system` property is that with `system` property, we need to set it but with custom lookup plugin, Log4j2 will load the value during initialization. However, to implement it we need to define a class which implements [`StrLookup`](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/lookup/StrLookup.html) interface. Probably, we can user our `LogFileDefiner` class to implement this interface and implement two abstract methods. It is similar to what was done for Logback. 

If we want to configure programmatically, we can initialize Log4j2 [combining xml configuration and programmatic configuration](https://logging.apache.org/log4j/2.x/manual/customconfig.html#Initialize_Log4j_by_Combining_Configuration_File_with_Programmatic_Configuration). Or [remove the appender and add a new one to root logger](https://logging.apache.org/log4j/2.x/manual/customconfig.html#Programmatically_Modifying_the_Current_Configuration_after_Initialization) which doesn't seem right to me.

To me using a custom lookup plugin seem a right approach. I think I have not missed anything. But yes, let me know if there are other approaches which I am not aware of. I remember, I spent most time to figure this out.

Reply:
Hey. Is this you? https://www.linkedin.com/in/badalsarkar/

Reply:
Yes. That's me. :-)

Reply:
Lets connect in LinkedIn.

Reply:
Done ;-)

Reply:
Implementation done and is working as expected. I will soon push it.